WooCommerce/Classes/POS/Presentation/PointOfSaleEntryPointView.swift

```diff
@@ -1,21 +1,15 @@
 import SwiftUI
 
 struct PointOfSaleEntryPointView: View {
-    // TODO:
-    // Temporary. DI proper product models once we have a data layer
-    private let viewModel: PointOfSaleDashboardViewModel = {
-        let products = POSProductFactory.makeFakeProducts()
-        let viewModel = PointOfSaleDashboardViewModel(products: products, cardReaderConnectionViewModel: .init(state: .scanningForReader(cancel: {})))
-
-        return viewModel
-    }()
+    @ObservedObject private var viewModel: PointOfSaleDashboardViewModel
```

Comment:
Just wondering, would `@StateObject` here fix the issue? Personally, I feel like owning the view model within the POS boundary (as if an invisible framework) would be preferable, and this seems to be the goal of `StateObject` in [the doc](https://developer.apple.com/documentation/swiftui/stateobject). The dependencies can be DI'ed through the view's initializer, and the view just does a simple initialization of the view model with the depenencies.

`@ObservedObject` view model works in the HubMenu because it's being held by the UIKit hosting controller which doesn't get recreated like in a SwiftUI view.

If `@StateObject` doesn't work out and the POS view model needs to be DI'ed to the POS view, maybe having `HubMenuViewModel` own it would be more consistent with the existing setup like `InPersonPaymentsMenuViewModel`.

Reply:
> Just wondering, would @StateObject here fix the issue? Personally, I feel like owning the view model within the POS boundary (as if an invisible framework) would be preferable, and this seems to be the goal of StateObject in [the doc](https://developer.apple.com/documentation/swiftui/stateobject). The dependencies can be DI'ed through the view's initializer, and the view just does a simple initialization of the view model with the depenencies.

You mean the View declaring its own dependencies via a `@StateObject` rather than DI it from outside? So:
```suggestion
    @StateObject private var viewModel = PointOfSaleDashboardViewModel(products: POSProductFactory.makeFakeProducts(),
                                                                       cardReaderConnectionViewModel: .init(state: .connectingToReader))
```
It does seem to fix the problem as well, I'm guessing because of:

> SwiftUI creates a new instance of the model object only once during the lifetime of the container that declares the state object. For example, SwiftUI doesn’t create a new instance if a view’s inputs change, but does create a new instance if the identity of a view changes

I like the idea of setting a POS boundary via using StateObject, and the benefits of dependency injection for testing purposes are not lost in this case since is just the initial pos state. I don't really know if there's any other trade-offs we should be aware of?

Reply:
Yup could be just `@StateObject private var viewModel: PointOfSaleDashboardViewModel` and setting the view model in the view init given the dependencies.

> I don't really know if there's any other trade-offs we should be aware of?

Good point, I can look more into this tomorrow. We've been using `@StateObject` in the code base in views that own a view model throughout the UI's lifetime.

Reply:
After some more reading on the official doc and other posts, `StateObject` feels to me like the most appropriate place for the view to own the view model that's only created once. With a few notes:

- The view model's dependencies shouldn't change, which I think is the case for POS - site ID, some data layer service implementation, analytics implementation (singleton in the app), etc.

> Explicit state object initialization works well when the external data that the object depends on doesn’t change for a given instance of the object’s container.

If we do want the view model to be recreated again when some dependency changes, it looks like we can achieve this by [binding the identity of the view](https://developer.apple.com/documentation/swiftui/stateobject#Force-reinitialization-by-changing-view-identity) that owns the StateObject view model to one or more parameters.

- StateObject should be private in all cases, and devs need to be aware not to re-initialize it again after the first initialization because any following inits wouldn't work silently and could result in unexpected behavior

---

Lemme know what you think!